### DISCUSSION TITLE
Controlnet Sketch and Scribble not working as intended?

### DISCUSSION THREAD

**Discussion by FreeeFry** [lllyasviel/stable-diffusion-webui-forge#310] [General]
I'll start this as a discussion instead of a bug report since I don't know if this is the intended behavior or not;

It seems that in forge with the integrated Controlnet both the Sketch and Scribble preprocessors aren't working the same way they do in regular a1111 webui in that they only seem to accept uploaded images and not support drawing manually on the canvas neither by creating a blank canvas and drawing on it or uploading an image and drawing on top of the image.

Is anyone else experiencing this?

For example
Using cyberrealistic_v41BackToBasics checkpoint with prompt `a large grassy field, a house in the center of a grassy field, realistic` Neg: `art, 3d, cartoon`and seed 4272455667 with the following Controlnet scribble:
![image](https://github.com/lllyasviel/stable-diffusion-webui-forge/assets/9145118/83b61479-03e8-4c42-bedb-603acbc9d9a8)

The image produced is different than without controlnet enabled but doesn't at all resemble the shape scribbled:

With Controlnet:
![tmpkyrbxwbb](https://github.com/lllyasviel/stable-diffusion-webui-forge/assets/9145118/46c723e1-45ea-4415-b2e4-33ee51497cf0)

Without:
![tmp7gkny79m](https://github.com/lllyasviel/stable-diffusion-webui-forge/assets/9145118/afbb27e6-6f41-498c-8bcc-48e73cb3f92e)

The scribbled image is shown as a separate image though in the output so it's at least being intercepted from the Controlnet input it seems;
![image](https://github.com/lllyasviel/stable-diffusion-webui-forge/assets/9145118/678a4acc-a9ef-4561-8b0b-65aec7cd78d6)

**Comment by FreeeFry** [comment]
If I save the Scribbled output image and load the image into Controlnet it works, so it seems that drawing on the canvas itself specifically is what is not working properly?

This is how the generated images looks with the scribble being uploaded as an image instead:
![tmpv7wa7bv3](https://github.com/lllyasviel/stable-diffusion-webui-forge/assets/9145118/4ad619f4-bc67-41fd-87fe-74bec520a2e1)
![tmp0v78v2vs](https://github.com/lllyasviel/stable-diffusion-webui-forge/assets/9145118/719dcb96-cf4f-4de3-b9fe-6e6727e883ac)

**Comment by cscooper2000** [comment]
I'm having the same issue.  In a1111 my workflow consisted of rendering a decent image and then loading it into the ControlNet (scribble) as a "reference image" (not sure of the terminology).  Now I could scratch in a few specific details in any area of the photo and like magic, I could create the exact details I intended.  

But, yeah... In Forge it doesn't seem to work at all.  I can still load an image, but tracing or sketching in new details has zero effect.  :(

**Comment by ostap667inbox** [comment]
That's your mistake, not ControlNet's :)
By drawing a sketch sketch on a black background you are essentially doing the work of a preprocessor. That is, you MUST select 'none' in the 'preprocessor' field. In this case, the model will work directly with your sketch. Instead, you are trying to get the preprocessor to make its own sketch from your sketch. Of course, nothing productive will come of it.
A preprocessor is needed when you use a photo, or other full-fledged image, as your source image.

**Reply by FreeeFry** [reply]
No, it is how it used to work (and afaik. still works) in the original implementation for automatic1111's webui, the new behavior is different and arguably a lot worse as it means you have to do any sketching using a reference image outside in an external editor and save that to an image which imho. defeats most of its use. @ostap667inbox 

I really hope this is not actually intended behavior..

**Reply by ostap667inbox** [reply]
The canvas in built-in ControlNet has a bug, you're right. It is not passed to the preview. But instead you can draw directly in the browser in the preview window. In Forge, the Photopea script is built into ControlNet:
![image](https://github.com/lllyasviel/stable-diffusion-webui-forge/assets/123593542/f8b413a2-f107-4a02-9c1f-b41f8b4d8dd8)

This solution definitely works:
![image](https://github.com/lllyasviel/stable-diffusion-webui-forge/assets/123593542/c5e863a9-b9fc-4c83-95c5-6bdde296bdd4)

It's strange that your screen doesn't show this button. Maybe photopea is disabled in your settings?
![image](https://github.com/lllyasviel/stable-diffusion-webui-forge/assets/123593542/204660f0-de9d-4f17-a6bf-81d919550155)

**Reply by FreeeFry** [reply]
I had disabled Photopea as I didn't like having my local webui connect to an external service.

I re-enabled it now to test this, but it doesn't work for me in Firefox at least, none of the edits I make are being sent back into the preprocessor preview. I feel it's a bit beside the point though as the original workflow that worked perfectly in a1111 was to simply upload a reference image into the preprocessor, draw on top of it and use the underlying reference as guide for details, etc.

I should perhaps have used an actual example in my initial post instead of drawing on top of a blank canvas, but imagine instead those drawings being on top of a reference image such as that of a grassy field or some such.

**Reply by FreeeFry** [reply]
I booted up and updated a1111 to latest version and retried this again and yeah it works pretty much as expected.

Poor drawing skills aside, this is a pretty nice workflow imho.
Sketch on top of a pre-rendered reference:
![image](https://github.com/lllyasviel/stable-diffusion-webui-forge/assets/9145118/c165d4b7-5c3c-4f58-9c6f-2fdf787db6f5)

Scribble pidinet processed:
![image](https://github.com/lllyasviel/stable-diffusion-webui-forge/assets/9145118/d225cbe9-f632-4a4d-ac2c-f5d59e0ffc6e)
Cherry picked image out of 4:
![image](https://github.com/lllyasviel/stable-diffusion-webui-forge/assets/9145118/70af1c8b-cb4f-4eb0-91fa-0ba8a945b075)

**Comment by makehillflat** [comment]
Before responding that someone is incorrect, it might be beneficial to review the question more thoroughly. Are you referring to Forge, or Auto1111? According to the original poster (OP), the 'sketch (in place)' function seems to be missing from the Forge controlnet implementation, and my observations confirm this. The option to select 'none' in the 'preprocessor' field does not appear to resolve this issue, nor does any setting that allows for sketch input. It would be more constructive to focus on understanding and addressing the actual issue at hand rather than seeking to correct others. If the solution is unclear, perhaps further investigation or assistance from others could be useful.

**Reply by ostap667inbox** [reply]
I suggested another solution above. Sorry, misled and triggered by this:
![scribble](https://github.com/lllyasviel/stable-diffusion-webui-forge/assets/123593542/eca0b33b-d582-45c9-ac76-85e758f7ce3d)

**Comment by cscooper2000** [comment]
Just a bit off-topic... You mentioned Photopea.  I was unaware of that until now.  You say it uses an internet connection to send data to their service?  But I'm reading otherwise online.  Any clarification around this?

Either way, it looks pretty cool.

**Reply by FreeeFry** [reply]
It tells you when you try to use it;
> `you are about to connect to https://photopea.com`
